src/protections/MentionFlood.ts

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2020 Emi Tatsuo Simpson et al.
+Copyright 2022 Marcel Radzio
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { Protection } from "./IProtection";
+import { Mjolnir } from "../Mjolnir";
+import { LogLevel, LogService } from "matrix-bot-sdk";
+import config from "../config";
+import { htmlEscape, isTrueJoinEvent } from "../utils";
+import { BooleanProtectionSetting, NumberProtectionSetting, OptionListProtectionSetting } from "./ProtectionSettings";
+
+const DEFAULT_MINUTES_BEFORE_TRUSTING = 20;
+const DEFAULT_MAX_MENTIONS_PER_MESSAGE = 20;
+const DEFAULT_REDACT = true;
+const DEFAULT_ACTION = "ban";
+
+const LOCALPART_REGEX = "[0-9a-z-.=_/]+";
+// https://github.com/johno/domain-regex/blob/8a6984c8fa1fe8481a4b99be0fa7f2a01ee17517/index.js
+const DOMAIN_REGEX = "(\\b((?=[a-z0-9-]{1,63}\\.)(xn--)?[a-z0-9]+(-[a-z0-9]+)*\\.)+[a-z]{2,63}\\b)";
+// https://stackoverflow.com/a/5284410
+const IPV4_REGEX = "(((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?))";
+// https://stackoverflow.com/a/17871737
+const IPV6_REGEX = "(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))";
+const PORT_REGEX = "(:[0-9]+)?";
+
+
+export class MentionFlood extends Protection {
+
+    settings = {
+        minutesBeforeTrusting: new NumberProtectionSetting(DEFAULT_MINUTES_BEFORE_TRUSTING),
+        maxMentionsPerMessage: new NumberProtectionSetting(DEFAULT_MAX_MENTIONS_PER_MESSAGE),
+        redact: new BooleanProtectionSetting(DEFAULT_REDACT),
+        action: new OptionListProtectionSetting(["ban", "kick", "warn"])
+    };
+
+    private justJoined: Map<string, Map<string, Date>> = new Map<string, Map<string, Date>>();
```

Comment:
FWIW, we now have `mjolnir.roomJoins.getUserJoin` that implements this feature in such a manner that all protections can use it.

Ideally, I'd prefer if you could use this instead of `justJoined`, as I'm a bit afraid it's going to grow very large for rooms such as MatrixHQ, so I'd like to avoid having two copies of this in memory. Do you feel that you have time to work on this? Otherwise, this can happen in a followup issue.